[#44][Tag] 태그 카프카 구현

leedongha1998 opened the pull request:
## Description
- 카프카 관련 설정
- 재시도, dql
- 테스트 코드

Resolves: #44 

## PR Type
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

Commit:
[#44][Tag] feature : 태그 쓰레드 구현

Commit:
Merge branch 'develop' of https://github.com/NewPick4U/NewPick4U into feature/tag/kafka

dannyseo9202 commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/config/KafkaConfig.java:
```diff
@@ -0,0 +1,99 @@
+  public DefaultErrorHandler errorHandler(KafkaTemplate<String, String> kafkaTemplate) {
+    DeadLetterPublishingRecoverer recoverer = new DeadLetterPublishingRecoverer(kafkaTemplate,
+        (record, ex) -> new TopicPartition(TOPIC_DLQ_TOPIC_NAME,
+            record.partition())); // 재시도 실패한 메세지를 dlq 토픽으로 전송
```
동하님 해당 부분의 의미는 "원래 메시지를 전송했던 파티션 번호와 동일한 번호의 파티션으로 DLQ 메시지를 전송한다"
의 의미입니다. 이에 따라서 아래와 같은 문제가 발생할 수 있습니다.

원본 토픽은 6개의 파티션이고, DLQ토픽은 1개의 파티션인 경우
원본 토픽 파티션 partition 0은 DLQ에서 동일하기 partion 0을 활용하게 됩니다.

하지만, DLQ의 파티션 갯수가 원본 토픽 파티션 보다 적은 경우 아래와 같은 문제가 발생할 수 있습니다.
원본 토픽 파티션 partition 5을 사용했기 때문에 DLQ에서도 partion 5를 사용하려 하지만 DLQ에는 partition 0 만 존재하는 경우.

이에 따라 해결 방법으로는 DLQ 토픽의 파티션 수도 원본 토픽 파티션 수로 같게 지정하거나,
위의 코드에서 파티션을 null로 설정하셔도 됩니다.
(record, ex) -> null (카프카가 라운드 로빙 방식으로 파티션을 자동 지정하기 떄문에 순서 보장이 안될 수 있습니다.)

dannyseo9202 commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/TagConsumer.java:
```diff
@@ -0,0 +1,67 @@
+
+      tagMessageHandler.handle(dto);
+
+      kafkaTemplate.send(TAG_TOPIC_NAME, record.value());
```
컨슈머 쪽에서 위와 같이 이벤트 발행하셔도 무방합니다 ~! (_ _) !

dannyseo9202 reviewed (commented):
수고하셨습니다 동하님 ~! :)

Commit:
[#44][Tag] refactor : KafkaConfig 의 dql 파티션 수정

Commit:
[#44][Tag] refactor : KafkaConfig 의 dql 파티션 수정 2

venus-y commented on tag/src/main/java/com/newpick4u/tag/application/usecase/TagServiceImpl.java:
```diff
@@ -36,4 +40,39 @@ public UpdateTagRequestDto updateTag(UpdateTagRequestDto tag, UUID tagId) {
+        Tag existingTag = tag.get();
+        existingTag.increaseScore();
+      } else {
+        log.info("태그 저장 시도 중~~");
```
if문 마지막에서 return으로 종료시키면 else를 생략할 수 있을 것 같습니다

leedongha1998 replied:
forEach 로 반복 돌리고 있는데 분기처리 안하고 return;으로 종료시키면 반복문 자체가 종료되지 않나요?

venus-y commented on tag/src/main/java/com/newpick4u/tag/application/usecase/TagServiceImpl.java:
```diff
@@ -36,4 +40,39 @@ public UpdateTagRequestDto updateTag(UpdateTagRequestDto tag, UUID tagId) {
+  public void deleteTagFromAi(AiNewsDto dto) {
+    dto.tagList().forEach(tagName -> {
+      Optional<Tag> optionalTag = tagRepository.findByTagName(tagName);
+
```
findByTagName의 결과에서 .orElseThrow로 값이 있을 경우 엔티티를 추출하고 그렇지 않을 경우 예외를 발생시키는
방향으로 처리해보면 optional.get을 피해볼 수 있을 것 같아보입니다!

venus-y commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/TagDlqConsumer.java:
```diff
@@ -0,0 +1,22 @@
+  public void hadleTagDlq(ConsumerRecord<String, String> record) {
+
+    log.error("[DLQ] 처리 실패한 메세지 수신 : {}", record.value());
+
```
이 부분도 수동커밋으로 작동한다면 acknowledge로 정상처리되었음을 확인하는 부분이 필요할 것 같습니다.

leedongha1998 replied:
넹 추가하겠습니다

venus-y replied:
동하님 여기서도 메세지 처리 성공, 실패를 분기할 수 있도록 try catch문이 필요할 것 같습니다.

venus-y commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/config/KafkaConfig.java:
```diff
@@ -0,0 +1,94 @@
+      log.warn("Retry #{} for record with key: {} due to {}", deliveryAttempt, record.key(),
+          ex.getMessage()); // 재시도 로그
+    });
+    return handler;
```
로깅하는 부분을 따로 빼주신 게 좋은 것 같습니다:)

venus-y commented on tag/src/test/java/com/newpick4u/tag/TagApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+
+  @Test
+  void contextLoads() {
+  }
```
저 같은 경우에는 이 테스트 클래스가 있을 경우 gradle build 시 카프카 의존성이 없어서 그런지
빌드가 터졌던 것 같습니다. 확인해보시고 저와 같은 현상이 발생할 경우 해당 클래스는 지워주셔도 무방할 것 같습니다.

venus-y commented on tag/src/test/java/com/newpick4u/tag/infrastructure/kafka/TagConsumerTest.java:
```diff
@@ -0,0 +1,108 @@
+    await()
+        .atMost(Duration.ofSeconds(60))      // 최대 5초 기다림
+        .pollInterval(Duration.ofMillis(1000)) // 100ms 마다 검사
+        .untilAsserted(() -> {
```
주석으로 달아주신 부분과 실제 매개변수로 들어간 수치가 달라보이는데 이 부분 한번 확인해주시면 좋을 것 같습니다!

venus-y reviewed (commented):
전체적으로 복잡한 로직이 많은데 잘 작성해주신 것 같습니다.
고생하셨습니다!

Commit:
[#44][Tag] refactor : 피드백 수정
- test 코드 주석 변경
- TagServiceImpl의 createTagFromAi 메소드 else 변경
- TagServiceImpl의 deleteTagFromAi 메소드 내 findByTagName : Optional -> tag 로 변경
- TagApplicationTests.java 삭제

Commit:
[#44][Tag] refactor : 테스트 수정
- KafkaConfig import 삭제
- test.yml : env file import

Commit:
Merge branch 'develop' of https://github.com/NewPick4U/NewPick4U into feature/tag/kafka

Commit:
[#44][Tag] refactor : dlqConsumer try catch 문 추가

Commit:
[#44][Tag] refactor : develop.yml 수정

Commit:
[#44][Tag] refactor : 태그 엔티티 수정
- tagId -> id

venus-y commented on tag/src/main/java/com/newpick4u/tag/application/usecase/TagMessageHandler.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
```
실질적인 비즈니스 로직을 처리하는 클래스는 태그서비스가 되어야 할 것으로 보이는데, `@Service`도 괜찮지만
`@Component` 어노테이션도 적절해보입니다.

venus-y commented on tag/src/main/java/com/newpick4u/tag/application/usecase/TagServiceImpl.java:
```diff
@@ -36,4 +40,41 @@ public UpdateTagRequestDto updateTag(UpdateTagRequestDto tag, UUID tagId) {
+    dto.tagList().forEach(tagName -> {
+      log.info("태그 조회 시도 중~~");
+      log.info("태그명 : {}", tagName);
+      Optional<Tag> tag = tagRepository.findByTagName(tagName); // 존재하는 태그는 score +1 처리
```
테스트성 로그는 추후에 지워주시면 좋을 것 같습니다

venus-y commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/TagConsumer.java:
```diff
@@ -0,0 +1,67 @@
+
+  @Value("${spring.kafka.tag-topic.topic-name}")
+  private String TAG_TOPIC_NAME;
+
```
상수로 정의해주신 것  같지는 않은데 대문자와 언더바를 혼합한 변수명을 사용해주신 이유를 알 수 있을까요??

venus-y reviewed (approved):
고생하셨습니다:)

Commit:
[#44][Tag] refactor : 피드백 수정 및 dlq 수정

Commit:
Merge branch 'develop' of https://github.com/NewPick4U/NewPick4U into feature/tag/kafka

hp5234 commented on tag/src/main/java/com/newpick4u/tag/domain/entity/Tag.java:
```diff
@@ -21,8 +21,8 @@ public class Tag extends BaseEntity {
-  @Column(name = "tag_id", columnDefinition = "CHAR(36)")
-  private UUID tagId;
+  @Column(name = "id", columnDefinition = "CHAR(36)")
+  private UUID id;
```
컨벤션 확인이 필요할것같습니다. 
엔티티 필드는 id 
테이블의 id 컬럼은 {테이블}_id 로 적용되어야할것같습니다.

hp5234 commented on tag/src/main/java/com/newpick4u/tag/infrastructure/kafka/TagConsumer.java:
```diff
@@ -0,0 +1,67 @@
+  )
+  public void consume(ConsumerRecord<String, String> record, Acknowledgment ack)
+      throws JsonProcessingException {
+    AiNewsDto dto = objectMapper.readValue(record.value(), AiNewsDto.class);
```
질문입니다. Json 처리에 실패해서 JsonProcessingException 가 throw 된다면 어떻게 동작하나요?

leedongha1998 replied:
예외가 발생하면 메세지는 수동 커밋이 되지않고 카프카 설정 클래스에서 예외를 잡아서 재시도 처리를 하게 됩니다.

Commit:
[#44][Tag] refactor : 태그 엔티티 컨벤션 수정

hp5234 reviewed (approved):
고생하셨습니다.
